### PR TITLE
Add clipboard service to copy text to user's clipboard

### PR DESCRIPTION
**Pull request by rohithbalaji123**
### Description
Clipboard service enabling JavaScript to copy text into user's clipboard.

### CODE REVIEW

File: frontend/src/service/utilities/Clipboard.service.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    textArea.value = text;
+
+    document.body.appendChild(textArea);
+    textArea.focus();
+    textArea.select();
+
+    let successful = false;
+    try {
+      successful = document.execCommand('copy');
+    } catch (err) {}
```
Comment (byliuyang) [inline frontend/src/service/utilities/Clipboard.service.ts]:
```suggestion
    } catch (err) {
         console.log(`Failed to copy ${text} into Clipboard`);
    }
```

File: frontend/src/service/utilities/Clipboard.service.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    return successful;
+  }
+
+  public copyTextToClipboard(text: string): Promise<void> {
+    if (!navigator.clipboard) {
+      console.log('other');
+      return this.fallbackCopyTextToClipboard(text)
+        ? Promise.resolve()
+        : Promise.reject();
+    }
```
Comment (byliuyang) [inline frontend/src/service/utilities/Clipboard.service.ts]:
```suggestion
    if (navigator.clipboard) {
      return navigator.clipboard.writeText(text);
    }
    const isSuccessful =  this.fallbackCopyTextToClipboard(text);
    if (isSuccessful) {
        return Promise.resolve();
    }
    return Promise.reject();
```

File: frontend/src/service/utilities/Clipboard.service.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    textArea.style.width = '2em';
+    textArea.style.height = '2em';
+
+    textArea.style.padding = '0';
+    textArea.style.border = 'none';
+    textArea.style.outline = 'none';
+    textArea.style.boxShadow = 'none';
+    textArea.style.background = 'transparent';
+
+    textArea.value = text;
```
Comment (byliuyang) [inline frontend/src/service/utilities/Clipboard.service.ts]: How about doing this instead?
textArea.style.display = 'none';

Comment (rohithbalaji123) [reply frontend/src/service/utilities/Clipboard.service.ts]: As per this method, the component should be there in the page to be able to select the text and copy to the clipboard. So, we are adding the component to the page, selecting, copying it and removing it from the page. Setting display to none won't let us select the text inside the textArea. The answer [here](https://stackoverflow.com/questions/400212/how-do-i-copy-to-the-clipboard-in-javascript) explains it in detail.

Comment (rohithbalaji123) [reply frontend/src/service/utilities/Clipboard.service.ts]: I tested changing the style to `textArea.style.display = 'none';` as well. It didn't work.

File: frontend/src/service/utilities/Clipboard.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    textArea.style.left = '0';
+
+    textArea.style.width = '2em';
+    textArea.style.height = '2em';
+
+    textArea.style.padding = '0';
+    textArea.style.border = 'none';
+    textArea.style.outline = 'none';
+    textArea.style.boxShadow = 'none';
+    textArea.style.background = 'transparent';
```
Comment (byliuyang) [inline frontend/src/service/utilities/Clipboard.service.ts]: Can this be wrapped within a private function called `createInvisibleTextArea()`?

Comment (rohithbalaji123) [reply frontend/src/service/utilities/Clipboard.service.ts]: Okay, this looks good.

File: frontend/src/service/clipboardService/Clipboard.factory.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { IClipboardService } from './Clipboard.service';
+import { NativeClipboardService } from './NativeClipboard.service';
+import { SimulatedClipboardService } from './SimulatedClipboard.service';
+
+export class ClipboardFactory {
```
Comment (byliuyang) [inline frontend/src/service/clipboardService/Clipboard.factory.ts]:
```suggestion
export class ClipboardServiceFactory {
```

Comment (byliuyang) [reply frontend/src/service/clipboardService/Clipboard.factory.ts]: This file can be renamed to `Clipboard.service.factory.ts`

File: frontend/src/service/clipboardService/Clipboard.factory.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  static createBrowserExtensionService(): IClipboardService {
```
Comment (byliuyang) [inline frontend/src/service/clipboardService/Clipboard.factory.ts]:
```suggestion
 public makeClipboardService(): IClipboardService {
```

Comment (byliuyang) [reply frontend/src/service/clipboardService/Clipboard.factory.ts]: We want to be able to swap out `ClipboardServiceFactory` implementation during testing using dependency injection. So we can't have static function here.

File: frontend/src/service/clipboardService/SimulatedClipboard.service.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { IClipboardService } from './Clipboard.service';
+
+export class SimulatedClipboardService implements IClipboardService {
+  private setupTextAreaWithText(text: string): HTMLTextAreaElement {
```
Comment (byliuyang) [inline frontend/src/service/clipboardService/SimulatedClipboard.service.ts]:
```suggestion
  private addInvisibleTextArea(text: string): HTMLTextAreaElement {
```

File: frontend/src/service/clipboardService/SimulatedClipboard.service.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    return textArea;
+  }
+
+  private selectTextInTextArea(textArea: HTMLTextAreaElement): void {
+    textArea.focus();
+    textArea.select();
+  }
+
+  private tearDownTextArea(textArea: HTMLTextAreaElement): void {
```
Comment (byliuyang) [inline frontend/src/service/clipboardService/SimulatedClipboard.service.ts]:
```suggestion
  private removeTextArea(textArea: HTMLTextAreaElement): void {
```

**Pull request merged**